chromadb/utils/batch_utils.py

```diff
@@ -1,36 +1,64 @@
-from typing import Optional, Tuple, List
-from chromadb.api import BaseAPI
-from chromadb.api.types import (
-    Documents,
-    Embeddings,
-    IDs,
-    Metadatas,
-)
+from typing import Any, Iterator, List, Optional, Tuple, TypeVar, Union, cast
+from chromadb.api import ClientAPI
+import numpy as np
+from rich.progress import track
+
+# (Can't provide type arguments to ndarray < 3.9.)
+T = TypeVar("T", bound=Tuple[Union[List[Any], np.ndarray, None], ...])  # type: ignore[type-arg]
 
 
 def create_batches(
-    api: BaseAPI,
-    ids: IDs,
-    embeddings: Optional[Embeddings] = None,
-    metadatas: Optional[Metadatas] = None,
-    documents: Optional[Documents] = None,
-) -> List[Tuple[IDs, Embeddings, Optional[Metadatas], Optional[Documents]]]:
-    _batches: List[
-        Tuple[IDs, Embeddings, Optional[Metadatas], Optional[Documents]]
-    ] = []
-    if len(ids) > api.get_max_batch_size():
-        # create split batches
-        for i in range(0, len(ids), api.get_max_batch_size()):
-            _batches.append(
-                (  # type: ignore
-                    ids[i : i + api.get_max_batch_size()],
-                    embeddings[i : i + api.get_max_batch_size()]
-                    if embeddings
-                    else None,
-                    metadatas[i : i + api.get_max_batch_size()] if metadatas else None,
-                    documents[i : i + api.get_max_batch_size()] if documents else None,
-                )
-            )
-    else:
-        _batches.append((ids, embeddings, metadatas, documents))  # type: ignore
-    return _batches
+    client: ClientAPI,
+    records: T,
+    max_batch_size: Optional[int] = 1024,
+    print_progress_description: Optional[str] = None,
+) -> Iterator[T]:
+    """
+    Takes tuples like `([0, 1], [2, 3])` and yields batches of the tuple like `([0], [2])` and `([1], [3])`.
+
+    For example:
+
+    ```python
+    import chromadb
+    from chromadb.utils.batch_utils import create_batches
+    import numpy as np
+
+    client = chromadb.Client()
+    collection = client.create_collection("foo")
+
+    ids = [str(i) for i in range(100_000)]
+    embeddings = np.random.rand(100_000, 128)
+
+    for (ids, embeddings) in create_batches(client, (ids, embeddings), print_progress_description="Adding documents..."):
+        collection.add(ids=ids, embeddings=embeddings)
+    ```
+
+    Args:
+        client: A chromadb client
+        records: A tuple of lists or numpy arrays
+        max_batch_size: The maximum batch size to use, defaults to 1024
+        print_progress_description: If specified, a progress bar will be displayed with this description
+    """
+    max_batch_size = min(client.get_max_batch_size(), max_batch_size or 1024)
+
+    set_size = -1
+    for field in records:
+        if isinstance(field, list):
```

Comment:
This doesn't do what you expect if you e.g. pass a single embedding on its own.

Reply:
It also allows passing in fields which aren't `list` - what would we do with them when processing batches?